### PR TITLE
Add tests for DDP control flow models.

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47206 Add tests for DDP control flow models.**

As discussed offline with @pritamdamania87, add testing to ensure per-iteration and rank-dependent control flow works as expected in DDP with `find_unused_parameters=True`.

Differential Revision: [D24659901](https://our.internmc.facebook.com/intern/diff/D24659901/)

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -3992,3 +3992,93 @@ def forward(_self, input, expected_type):  # noqa
...
+        def test_ddp_control_flow_same_across_ranks(self):
+            # Control flow that is the same across ranks.
+            class ToyModel(nn.Module):
+                def __init__(self):
+                    super(ToyModel, self).__init__()
+                    self.lin1 = nn.Linear(10, 10, bias=False)
+                    self.lin2 = nn.Linear(10, 10, bias=False)
+                    self.counter = 0
+
+                def forward(self, x):
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:4009]: Ideally we should test input dependent control flows (conditionals on x).

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -3992,3 +3992,93 @@ def forward(_self, input, expected_type):  # noqa
...
+            # Control flow that is different across ranks.
+            class ToyModel(nn.Module):
+                def __init__(self, rank):
+                    super(ToyModel, self).__init__()
+                    self.lin1 = nn.Linear(10, 10, bias=False)
+                    self.lin2 = nn.Linear(10, 10, bias=False)
+                    self.counter = 0
+                    self.rank = rank
+
+                def forward(self, x):
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:4053]: same here, control flow should be on `x`

File: torch/testing/_internal/distributed/distributed_test.py [hunk 3]
```diff
@@ -3992,3 +3992,93 @@ def forward(_self, input, expected_type):  # noqa
...
+                    self.counter += 1
+                    # Unused parameter on even iterations for rank 1.
+                    if self.counter % 2 == 0 and self.rank == 1:
+                        return self.lin2(F.relu(self.lin1(x)))
+                    else:
+                        return F.relu(self.lin1(x))
+
+            torch.cuda.set_device(self.rank)
+            model = torch.nn.parallel.DistributedDataParallel(
+                ToyModel(self.rank).cuda(self.rank), device_ids=[self.rank], find_unused_parameters=True
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:4063]: Can we also add a test that verifies we raise an appropriate error when `find_unused_parameters=False`?

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py:4010]: Updated to test based on `x`, basically we do a `torch.equal` using x and an expected tensor. Please let me know if there is a better way to test that.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 3 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py]: Added this to both the tests.

**Review by pritamdamania87 — APPROVED** [review]

**Comment by zou3519** [comment]
The tests introduced by this PR are failing: https://app.circleci.com/pipelines/github/pytorch/pytorch/235338/workflows/76ab99d0-701b-447e-8a67-7b0cc26ba7b5/jobs/8713716 so I'm reverting this

**Comment by rohan-varma** [comment]
It looks like multigpu tests don't run on commits, and I tested only with a 2 GPU machine, which is why this was missed. Will resubmit with a ci-all.

**Pull request closed without merging**